### ISSUE TITLE
HTTP Code 404 lead to an HTTP 500 (Internal server error)

### ISSUE THREAD

**Issue by dwursteisen** [bclozel/spring-boot-web-reactive#22]
Hello, 

I try to use Spring Boot using the starter-web-reactive with reactor-netty.

If I try to generate a 404 http code, an error occur in the server which lead to an HTTP code 500 instead.
to achieve this, I try to use an Exception mapped with the 404 code, or using an `ResponseEntity` with the 404 code.

I found an work arround using an `ResponseEntity` with an empty body and the 404 code. Unfortunately, I haven't found why it doesn't work and another exception is throwed.

# How to get get this issue ?

- Use this bellow Spring Boot application
- start it
- open http://localhost:8080/notFound1 (error 500)
- open http://localhost:8080/notFound2 (error 500)
- open http://localhost:8080/notFound3 (error 404)

__Expected :__

- open http://localhost:8080/notFound1 (error 404)
- open http://localhost:8080/notFound2 (error 404)
- open http://localhost:8080/notFound3 (error 404)

```
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;

import rx.Single;

@SpringBootApplication
public class IssueApplication {

	@ResponseStatus(HttpStatus.NOT_FOUND)
	public static class NotFound extends RuntimeException {

	}

	@Controller
	public static class ExampleController {
		@RequestMapping("/notFound1")
		public Single<String> notFound1() {
			return Single.error(new NotFound());
		}

		@RequestMapping("/notFound2")
		public Single<ResponseEntity<String>> notFound2() {
			return Single.just(new ResponseEntity<>(HttpStatus.NOT_FOUND));
		}

		@RequestMapping("/notFound3")
		public Single<ResponseEntity<String>> notFound3() {
			return Single.just(new ResponseEntity<>("", HttpStatus.NOT_FOUND));
		}
	}

	public static void main(String[] args) {
		SpringApplication.run(IssueApplication.class, args);
	}
}
```

**Comment by bclozel** [comment]
Hi @dwursteisen - are you using Boot 1.4.0 or the latest SNAPSHOT?
I've recently [SPR-14618](https://jira.spring.io/browse/SPR-14618) which might help here.

**Comment by snicoll** [comment]
@dwursteisen generate another project with Spring Boot `1.4.1.BUILD-SNAPSHOT` and it will give you the snapshot starter (with the latest changes).

**Comment by dwursteisen** [comment]
I'm using spring-bbot-starter-parent with 1.4.1.BUILD-SNAPSHOT with reactor 3.0.0.BUILD-SNAPSHOT, spring 5.0.0.BUILD-SNAPSHOT and spring-boot-dependencies-web-reactive 0.1.0.BUILD-SNAPSHOT.

I just look more into this issue : I thinks it's two different issues.

With logs : 
```
logging.level.org.springframework=DEBUG
```

#### notFound1 

```
com.example.IssueApplication$NotFound: null
	at com.example.IssueApplication$ExampleController.notFound1(IssueApplication.java:24) ~[classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_65]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_65]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_65]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_65]
	at org.springframework.web.reactive.result.method.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:172) ~[spring-web-reactive-5.0.0.BUILD-20160902.094047-157.jar:5.0.0.BUILD-SNAPSHOT]
	at org.springframework.web.reactive.result.method.InvocableHandlerMethod.lambda$invokeForRequest$0(InvocableHandlerMethod.java:95) ~[spring-web-reactive-5.0.0.BUILD-20160902.094047-157.jar:5.0.0.BUILD-SNAPSHOT]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:145) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply.subscribe(MonoThenApply.java:50) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoMapFuseable.subscribe(MonoMapFuseable.java:66) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply$MonoThenApplyManager.onNext(MonoThenApply.java:131) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:71) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:81) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:383) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onSubscribe(FluxConcatMap.java:192) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:95) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:59) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxConcatMap.subscribe(FluxConcatMap.java:116) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoNext.subscribe(MonoNext.java:44) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwiseIfEmpty.subscribe(MonoOtherwiseIfEmpty.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply.subscribe(MonoThenApply.java:57) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply.subscribe(MonoThenApply.java:57) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenSupply$MonoConcatIgnoreManager.drain(MonoThenSupply.java:166) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenSupply.subscribe(MonoThenSupply.java:54) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.ipc.netty.http.NettyHttpServerHandler.channelRead(NettyHttpServerHandler.java:92) ~[reactor-netty-0.6.0.BUILD-20160902.020552-26.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:350) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:435) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:250) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:350) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:571) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:512) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:426) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:398) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:877) ~[netty-common-4.1.3.Final.jar:4.1.3.Final]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_65]
```
(This stack trace is not relevan to me)

#### notFound2

```
java.lang.NullPointerException: null
	at java.lang.Class.isAssignableFrom(Native Method) ~[na:1.8.0_65]
	at org.springframework.core.codec.ByteBufferEncoder.canEncode(ByteBufferEncoder.java:46) ~[spring-core-5.0.0.M1.jar:5.0.0.M1]
	at org.springframework.http.codec.EncoderHttpMessageWriter.canWrite(EncoderHttpMessageWriter.java:63) ~[spring-web-5.0.0.BUILD-20160902.094047-214.jar:5.0.0.BUILD-SNAPSHOT]
	at org.springframework.web.reactive.result.method.annotation.AbstractMessageWriterResultHandler.lambda$getProducibleMediaTypes$0(AbstractMessageWriterResultHandler.java:135) ~[spring-web-reactive-5.0.0.BUILD-20160902.094047-157.jar:5.0.0.BUILD-SNAPSHOT]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174) ~[na:1.8.0_65]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374) ~[na:1.8.0_65]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[na:1.8.0_65]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[na:1.8.0_65]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[na:1.8.0_65]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_65]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[na:1.8.0_65]
	at org.springframework.web.reactive.result.method.annotation.AbstractMessageWriterResultHandler.getProducibleMediaTypes(AbstractMessageWriterResultHandler.java:137) ~[spring-web-reactive-5.0.0.BUILD-20160902.094047-157.jar:5.0.0.BUILD-SNAPSHOT]
	at org.springframework.web.reactive.result.method.annotation.AbstractMessageWriterResultHandler.writeBody(AbstractMessageWriterResultHandler.java:113) ~[spring-web-reactive-5.0.0.BUILD-20160902.094047-157.jar:5.0.0.BUILD-SNAPSHOT]
	at org.springframework.web.reactive.result.method.annotation.ResponseEntityResultHandler.lambda$handleResult$2(ResponseEntityResultHandler.java:161) ~[spring-web-reactive-5.0.0.BUILD-20160902.094047-157.jar:5.0.0.BUILD-SNAPSHOT]
	at reactor.core.publisher.MonoThenApply$MonoThenApplyManager.onNext(MonoThenApply.java:97) [reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.adapter.RxJava1Adapter$SingleAsMono$SingleAsMonoSubscriber.onSuccess(RxJava1Adapter.java:436) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at rx.Single$18.onNext(Single.java:1797) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.observers.SafeSubscriber.onNext(SafeSubscriber.java:139) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.internal.producers.SingleDelayedProducer.emit(SingleDelayedProducer.java:102) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.internal.producers.SingleDelayedProducer.setValue(SingleDelayedProducer.java:85) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.Single$1$1.onSuccess(Single.java:84) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.internal.util.ScalarSynchronousSingle$1.call(ScalarSynchronousSingle.java:40) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.internal.util.ScalarSynchronousSingle$1.call(ScalarSynchronousSingle.java:36) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.Single$1.call(Single.java:94) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.Single$1.call(Single.java:74) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.Single.subscribe(Single.java:1725) ~[rxjava-1.1.5.jar:1.1.5]
	at rx.Single.subscribe(Single.java:1802) ~[rxjava-1.1.5.jar:1.1.5]
	at reactor.adapter.RxJava1Adapter$SingleAsMono.subscribe(RxJava1Adapter.java:398) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply.subscribe(MonoThenApply.java:57) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply$MonoThenApplyManager.onNext(MonoThenApply.java:131) [reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:913) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply$MonoThenApplyManager$SecondSubscriber.onNext(MonoThenApply.java:201) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxResume$ResumeSubscriber.onNext(FluxResume.java:74) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:126) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:1529) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:182) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:1246) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxResume$ResumeSubscriber.onSubscribe(FluxResume.java:69) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:96) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:169) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply.subscribe(MonoThenApply.java:50) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoMapFuseable.subscribe(MonoMapFuseable.java:66) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply$MonoThenApplyManager.onNext(MonoThenApply.java:131) [reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:71) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:81) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:383) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onSubscribe(FluxConcatMap.java:192) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:95) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:59) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.FluxConcatMap.subscribe(FluxConcatMap.java:116) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoNext.subscribe(MonoNext.java:44) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwiseIfEmpty.subscribe(MonoOtherwiseIfEmpty.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply.subscribe(MonoThenApply.java:57) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenApply.subscribe(MonoThenApply.java:57) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoOtherwise.subscribe(MonoOtherwise.java:47) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenSupply$MonoConcatIgnoreManager.drain(MonoThenSupply.java:166) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.core.publisher.MonoThenSupply.subscribe(MonoThenSupply.java:54) ~[reactor-core-3.0.0.BUILD-20160818.233648-83.jar:na]
	at reactor.ipc.netty.http.NettyHttpServerHandler.channelRead(NettyHttpServerHandler.java:92) ~[reactor-netty-0.6.0.BUILD-20160902.020552-26.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:350) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:435) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:250) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:350) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:571) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:512) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:426) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:398) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:877) ~[netty-common-4.1.3.Final.jar:4.1.3.Final]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_65]
```

which lead to this code:

```
   public boolean canEncode(ResolvableType elementType, MimeType mimeType, Object... hints) {
        Class clazz = elementType.getRawClass();
        return super.canEncode(elementType, mimeType, hints) && ByteBuffer.class.isAssignableFrom(clazz);
    }
```
At runtime, `clazz` is null. (got the same error using a `Mono` instead of a `Single`)


I really don't understand the first stacktrace. The second issue should be because the ResponseEntity hasn't a body, and it try still to encode it.

**Comment by dwursteisen** [comment]
@snicoll I just did. Same result

(I generate it from start.spring.io + 1.4.1-SNAPSHOT + Reactive Web. Then, after download, I just add RxJava 1.1.9 as dependencies)

```
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building demo 0.0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-dependency-plugin:2.10:tree (default-cli) @ demo ---
[INFO] com.example:demo:jar:0.0.1-SNAPSHOT
[INFO] +- org.springframework.boot.experimental:spring-boot-starter-web-reactive:jar:0.1.0.BUILD-SNAPSHOT:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter:jar:1.4.1.BUILD-SNAPSHOT:compile
[INFO] |  |  +- org.springframework.boot:spring-boot:jar:1.4.1.BUILD-SNAPSHOT:compile
[INFO] |  |  |  \- org.springframework:spring-context:jar:5.0.0.BUILD-SNAPSHOT:compile
[INFO] |  |  |     +- org.springframework:spring-aop:jar:5.0.0.BUILD-SNAPSHOT:compile
[INFO] |  |  |     +- org.springframework:spring-beans:jar:5.0.0.BUILD-SNAPSHOT:compile
[INFO] |  |  |     \- org.springframework:spring-expression:jar:5.0.0.BUILD-SNAPSHOT:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-autoconfigure:jar:1.4.1.BUILD-SNAPSHOT:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-starter-logging:jar:1.4.1.BUILD-SNAPSHOT:compile
[INFO] |  |  |  +- ch.qos.logback:logback-classic:jar:1.1.7:compile
[INFO] |  |  |  |  \- ch.qos.logback:logback-core:jar:1.1.7:compile
[INFO] |  |  |  +- org.slf4j:jcl-over-slf4j:jar:1.7.21:compile
[INFO] |  |  |  +- org.slf4j:jul-to-slf4j:jar:1.7.21:compile
[INFO] |  |  |  \- org.slf4j:log4j-over-slf4j:jar:1.7.21:compile
[INFO] |  |  \- org.yaml:snakeyaml:jar:1.17:runtime
[INFO] |  +- org.springframework.boot.experimental:spring-boot-autoconfigure-web-reactive:jar:0.1.0.BUILD-SNAPSHOT:compile
[INFO] |  |  +- org.springframework:spring-web-reactive:jar:5.0.0.BUILD-SNAPSHOT:compile
[INFO] |  |  |  +- io.projectreactor:reactor-core:jar:3.0.0.BUILD-SNAPSHOT:compile
[INFO] |  |  |  +- org.reactivestreams:reactive-streams:jar:1.0.0:compile
[INFO] |  |  |  \- org.springframework:spring-web:jar:5.0.0.BUILD-SNAPSHOT:compile
[INFO] |  |  \- io.netty:netty-buffer:jar:4.1.3.Final:compile
[INFO] |  |     \- io.netty:netty-common:jar:4.1.3.Final:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-tomcat:jar:1.4.1.BUILD-SNAPSHOT:compile
[INFO] |  |  +- org.apache.tomcat.embed:tomcat-embed-core:jar:8.5.4:compile
[INFO] |  |  +- org.apache.tomcat.embed:tomcat-embed-el:jar:8.5.4:compile
[INFO] |  |  \- org.apache.tomcat.embed:tomcat-embed-websocket:jar:8.5.4:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-databind:jar:2.7.6:compile
[INFO] |     +- com.fasterxml.jackson.core:jackson-annotations:jar:2.7.6:compile
[INFO] |     \- com.fasterxml.jackson.core:jackson-core:jar:2.7.6:compile
[INFO] +- org.springframework.boot:spring-boot-starter-test:jar:1.4.1.BUILD-SNAPSHOT:test
[INFO] |  +- org.springframework.boot:spring-boot-test:jar:1.4.1.BUILD-SNAPSHOT:test
[INFO] |  +- org.springframework.boot:spring-boot-test-autoconfigure:jar:1.4.1.BUILD-SNAPSHOT:test
[INFO] |  +- com.jayway.jsonpath:json-path:jar:2.2.0:test
[INFO] |  |  +- net.minidev:json-smart:jar:2.2.1:test
[INFO] |  |  |  \- net.minidev:accessors-smart:jar:1.1:test
[INFO] |  |  |     \- org.ow2.asm:asm:jar:5.0.3:test
[INFO] |  |  \- org.slf4j:slf4j-api:jar:1.7.21:compile
[INFO] |  +- junit:junit:jar:4.12:test
[INFO] |  +- org.assertj:assertj-core:jar:2.5.0:test
[INFO] |  +- org.mockito:mockito-core:jar:1.10.19:test
[INFO] |  |  \- org.objenesis:objenesis:jar:2.1:test
[INFO] |  +- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] |  +- org.hamcrest:hamcrest-library:jar:1.3:test
[INFO] |  +- org.skyscreamer:jsonassert:jar:1.3.0:test
[INFO] |  |  \- org.json:json:jar:20140107:test
[INFO] |  +- org.springframework:spring-core:jar:5.0.0.BUILD-SNAPSHOT:compile
[INFO] |  \- org.springframework:spring-test:jar:5.0.0.BUILD-SNAPSHOT:test
[INFO] \- io.reactivex:rxjava:jar:1.1.9:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.013 s
[INFO] Finished at: 2016-09-02T13:29:22+02:00
[INFO] Final Memory: 19M/309M
[INFO] ------------------------------------------------------------------------
```

**Comment by bclozel** [comment]
Thanks @dwursteisen , I've got enough information for now. Looking into this.

**Comment by bclozel** [comment]
I've created and fixed [SPR-14663](https://jira.spring.io/browse/SPR-14663) so the  `NullPointerException` in `"notFound2"` should disappear. The fix should be available shortly in the next Spring Framework 5.0.0.BUILD-SNAPSHOT.

About `"notFound1`, we don't support `@ResponseStatus` for now, but we're covering that feature in a more friendlier way for a reactive pipeline.

Here's how you can throw a `ResponseStatusException` (or something that inherits from it):

``` java
@RequestMapping("/notFound1")
public Mono<String> notFound1() {
  return Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND, "Resource not found"));
}
```

Now you can map business exceptions to those:

``` java
@RequestMapping("/notFound1")
public Mono<String> notFound1() {
  return Mono
    .just("hello")
    .mapError(MyCustomException.class,
        e -> new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage()));
}
```

Thanks for the report!
